Oceanwater 518 switch to using effort controllers joint trajectory controller

Samahu opened the pull request:
## Linked Issues
Linked PR: https://github.com/nasa/ow_autonomy/pull/14
Epic Link: [OCEANWATER-518 | Switch to using effort_controllers/JointTrajectoryController](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-518)
Child Issues:
- ~[OCEANWATER-483 | API Improvements: Remove delete_prev_traj parameter from all of the services](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-483)~
- ~[OCEANWATER-519 | Re-run moveit setup assistant with JointTrajectoryController](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-519)~
- ~[OCEANWATER-524 | Integrate the new JointTrajectoryController confg with OceanWATERS sim](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-524)~
- ~[OCEANWATER-525 | Refactor arm services to use moveit based on JointTrajectoryController interface (arm_controller)](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-525)~
- ~[OCEANWATER-536 | Refactor arm services to use moveit based on JointTrajectoryController interface (limbs_controller)](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-536)~
- ~[OCEANWATER-537 | Refactor arm services to use moveit based on JointTrajectoryController interface (grind_controller)](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-537)~
- ~[OCEANWATER-526 | Remove code path that perform trajectory recording and the trajectory publisher](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-526)~
- ~[OCEANWATER-527 | Add the ability to track the execution of a trajectory and stop it](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-527)~
- ~[OCEANWATER-528 | Tune PID controllers for the JointTrajectoryController](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-528)~
- ~[OCEANWATER-529 | Pick proper joint_trajectory_controller goal and trajectory constraints](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-529)~
- ~[OCEANWATER-530 | Enable switching between arm controllers at runtime](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-530)~
> - [OCEANWATER-531 | basic trajectory plans concatenation](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-531) [DESCOPED!]
> - [OCEANWATER-533 | Rework communication between the different move services and ow_autonomy](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-533) [DESCOPED!]
> - [OCEANWATER-534 | Implement the current arm services as rosaction](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-534) [DESCOPED!]
- ~[OCEANWATER-540 | Factor out ground detection functionality out of trajectory_publisher](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-540)~
- ~[OCEANWATER-541 | Fix disconnected frame tree](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-541)~
- ~[OCEANWATER-544 | Ensure RRStar planner is set during arm operations ](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-544)~
- ~[OCEANWATER-559 | Revise how antenna control fits within JointTrajectoryController](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-559)~
- ~[OCEANWATER-549 | Update ow_lander/README.md with new changes](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-549)~
- [OCEANWATER-550 | Update the arm simulation section within the user guide](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-550) [PENDING...]
- ~[OCEANWATER-567 | Remove all references to trajectory_publisher from ow_autonomy](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-567)~
> - [OCEANWATER-570 | Use arm_controller for performing the dig_circular activity](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-570) [DESCOPED!]
- [OCEANWATER-569 | RRTConnect: Unable to sample any valid states for goal tree](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-569) [PENDING...]

## Summary of Changes
* Re-ran the **moveit setup assistant** selecting the `effort_controllers/JointTrajectoryContoller` as the main controller for the different planning group:
  - Defined poses through setup assistant and use them through moveit
  - Address existing problems in the current config
  - Keep  the .setup_assistant file as it would help easily us revise things .. i.e. add new poses, change controllers, ...
* Refactored some of the arm services to bypass the trajectory recording functionality and rely completely on joint trajectory controller interface. Successfully ported services so far are:
  - **Stow**
  - **Unstow**
  - **GuardedMove**
  - **DeliverSample**
* **Stow** and **Unstow** services has been simplified by utilizing predefined poses.
* Added back the grinder_controller (deactivated by default)
* Added a new limbs_controller (deactivated by default)
* Created a method that enables switching between two sets of controllers at runtime
* Ported the remaining services to using _JointTrajectoryController_:
  - **DigCircular**
  - **DigLinear**
  - **Grind**
* Removed `delete_prev_traj` from these services definitions
* Removed deprecated code that performs trajectory recording
* Added the ability to track the execution of a trajectory and stop it (implemented in TrajectoryAsyncExecuter)
* Ground detection functionality moved out of `trajectory_publisher.py` to `path_commander_interface.py`
  - `trajectory_publisher.py` is completely dropped
* **Guarded Move** activity now publishes the result of ground detection as before
* Keep antenna operation and control as before (antenna not using JointTrajectoryController)
* Tuned the JointTrajectoryController trajectory and goal constraints
* Bring up `README.md` up to date with the new changes

## Test
Lauch the simulation using any of the provided launch files:
```bash
roslaunch ow (atacama_y1a.launch | europa_terminator.launch | europa_terminator_workspace.launch)
```
### Test the arm moveit configuration using RVIZ
Then use the associated rviz window to command the robot in RVIZ and Gazebo
![image](https://user-images.githubusercontent.com/606033/99743191-07399600-2a8a-11eb-95d7-5c3ee3a2a89d.png)

### Test ported services
Use either **rqt_service_caller** interface or launch a separate terminal (with the workspace sourced) and invoked any of the ported commands
```bash
rosservice call /arm/stow
rosservice call /arm/unstow
rosservice call /arm/deliver_sample ...
rosservice call /arm/guarded_move ...
rosservice call /arm/dig_linear ...
rosservice call /arm/dig_circular ...
rosservice call /arm/grind ...
```

### Testing Guarded Move
While the simulation is running, open another terminal (with the workspace sourced) use the following command to monitor the output of `/guarded_move_result` topic:
```bash
rostopic echo /guarded_move_result
```
Use the rqt_service_caller interface or using a separate terminal (with the workspace sourced) and invoked any of the ported commands
```bash
rosservice call /arm/guarded_move
```

kmdalal commented on ow_lander/CMakeLists.txt:
```diff
@@ -1,5 +1,4 @@
-cmake_minimum_required(VERSION 2.8.3)
-
+cmake_minimum_required(VERSION 3.1.3)
```
Should this be the min required version in all our packages now?

Samahu replied:
We can if we want to .. I think on our computers everyone has version 3.10

kmdalal replied:
If catkin 3.1.3 or newer is standard with ROS Melodic, this seems the right choice.  But if some of our users may be using an older catkin, 3.1 is more restrictive if not actually "required".

Samahu replied:
this is a cmake version, it is different than catkin. I think on Ubuntu 18.04 you get cmake 3.10 but don't quote me on it

mogumbo replied:
ROS melodic packages originally used cmake 2.8.3. Some of them (such as https://github.com/ros/ros_comm) have been updated to 3.0.2. noetic is also on 3.0.2. To keep things useable for people who haven't updated their computers, I think using either of these versions is okay. I don't see a need to go beyond 3.0.2 unless there's a specific feature we need.

Samahu replied:
This version bump came from moveit setup_assistant. I will try to lower it back and see if it breaks anything

Samahu replied:
addressed

mogumbo reviewed (changes requested):
Please update ow_lander/README.md with any changes you think are important to the Setup Assistant notes.

mogumbo commented on ow_lander/launch/spawn.launch:
```diff
@@ -71,22 +56,42 @@
+
+  <!-- Load the URDF, SRDF and other .yaml configuration files on the param server -->
+  <include file="$(find ow_lander)/launch/planning_context.launch">
+    <arg name="load_robot_description" value="false"/>
```
load_robot_description is used as a conditional on the param that parses the xacro file, but "if" is not a valid attribute for <param>: http://wiki.ros.org/roslaunch/XML/param. That <param> line ends up overriding the one here in spawn.launch, which causes the lander to be disconnected from the frame tree and causes the 3 worlds to be dark. Remove <param> xacro line from planning_context.launch to fix all this. This load_robot_description arg will be obsolete and can also be removed.

Samahu replied:
Oh I see, Great catch!

Samahu replied:
Actually, according to http://wiki.ros.org/roslaunch/XML#if_and_unless_attributes it says that if/unless are supported on every tag and I did verify that the *param if* is a valid construct. However, I suspect that the value $(arg load_robot_description) is either getting a 'true' value or not being honored!

Samahu replied:
Somehow it is getting a non-false value

Samahu replied:
I think I have an intuition at what's happening. The only other file that is being loaded and uses the  same arg/param is move_group.launch and that one sets default value to true but I thought it wouldn't override the value being set by spawn.launch ... still looking at it but that's really weird

Samahu replied:
Ok, so I have addressed it by setting **load_robot_description** to **false** when including `move_group.launch` within `spawn.launch`. I did verify that this doesn't break existing functionality.

Samahu replied:
Note that this change also addresses the ticket https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-524 (which was created to ensure the configuration works for _atacama_y1a_, _europa_terminator_workspace_, initially only the _europa_terminator_ world could be loaded)

Samahu commented on ow_lander/config/lander.srdf:
```diff
@@ -22,58 +22,96 @@
+    <disable_collisions link1="l_ant_foot" link2="l_dist" reason="Never" />
     <disable_collisions link1="l_ant_foot" link2="l_prox" reason="Never" />
     <disable_collisions link1="l_ant_foot" link2="l_shou" reason="Never" />
+    <disable_collisions link1="l_ant_foot" link2="lander_light0" reason="Never" />
```
I think it is possible to quantify this however would take time and effort to do so.
The setup_assistant should be re-run for this purpose (collision matrix) whenever we change links (count, size, shape), joints (count,type). I can add to the README.md.
The most important thing is re-running the setup-assistant is a lot easier now with `roslaunch ow_lander setup_assistant.launch`

Samahu replied:
I did add some about re-run the moveit setup assistant to the ow_lander/README.md. However, I want to note that there are plenty of not very useful information in that file. It should probably be overhauled after we are done with this epic! (I'll add a task for it)

Samahu commented on ow_lander/config/moveit.rviz:
```diff
@@ -66,7 +63,7 @@ Visualization Manager:
           Y: 2
           Z: 2
-      Name: MotionPlanning_ArmPlanner
+      Name: MotionPlanning
```
The previous name is misleading actually, this is the name of the plugin not the name of the planner.
The MotionPlanning plugin can be used to control the arm + antenna + any other defined joints/move group.

kmdalal commented on ow_lander/README.md:
````diff
@@ -65,6 +67,18 @@ arguments (Careful, the bools have a capital first letter: `True`, `False`),
+roslaunch ow_lander setup_assistant.launch
+``` 
+
+There are different reasons why would you may want to re-run the setup assistant:
````
'would you may' -> 'you would'   or 'you may'

Samahu replied:
Addressed

kmdalal commented on ow_lander/launch/moveit_rviz.launch:
```diff
@@ -1,17 +1,15 @@
 <launch>
 
   <arg name="debug" default="false" />
-  <arg name="use_sim_time" value="true"/>
```
Just curious, why was this removed?

Samahu replied:
The whole file should be deprecated actually!

Samahu replied:
This file is replaced by moveit_rviz.launch which is the default name used by moveit setup assistant

kmdalal replied:
Ah, I see that now.  Wasn't familiar with the file replacement notation in these PRs.

Samahu replied:
By default **use_sim_time** is not included in that file. For its value to have any effect, the argument needs to be propagated to empty_world.launch. So I need to trace if defining this argument in this file has any effect at all.

Samahu replied:
So on melodic-devel this argument has no purpose in this file, which means removing it is the proper thing

mogumbo replied:
Good catch. We definitely want use_sim_time = true to be set when we launch gazebo so that gazebo uses the date stamp we set in the world file.

kmdalal replied:
And Devin relied on this parameter for sim speedup, though I'm not sure how it got propagated from ow_autonomy launch files.

Samahu commented:
Whoever have time today to test this branch please do so. When doing so make sure that you update your workspace to OCEANWATER-518 on **ow_simulator** and **ow_autonomy**.

There are still two things that are pending for this PR:
* https://github.com/nasa/ow_simulator/pull/54
* also [OCEANWATER-528 | Tune PID controllers for the JointTrajectoryController](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-528) which will probably lag a bit.

If you start noticing too much "LCP error" message in the log, unfortunately these have happened after I synchronized this branch with *melodic-devel* to resolve merge conflicts before actually merging.

Samahu commented:
> If you start noticing too much "LCP error" message in the log, unfortunately these have happened after I synchronized this branch with _melodic-devel_ to resolve merge conflicts before actually merging.

The LCP error message are found in the branch before the synchronization with melodic-devel. I am not sure at what point of this branch/PR these started to happen. It wasn't much of a problem early on.

These are not necessarily an issue as we can silence them

kmdalal commented:
When I launch atacama, the antenna is shaking in both Gazebo and Rviz.  I presume this is a known and unrelated issue.

damianacatanoso commented:
> When I launch atacama, the antenna is shaking in both Gazebo and Rviz. I presume this is a known and unrelated issue.

@Samahu I would try to merge the latest lander.xacro. This has new values for the antenna parameters

kmdalal commented:
Rviz now keeps repeating the planning motion, which is pretty distracting.  Is there a way to stop it or have it just do the motion once?

damianacatanoso commented:
> Rviz now keeps repeating the planning motion, which is pretty distracting. Is there a way to stop it or have it just do the motion once?

There's a simple way to stop it, I had suggested this to Ussama once.

kmdalal commented:
Has anyone run an autonomy plan with this branch (ow_autonomy needs to be on the same branch).   Service calls are no longer working for me from autonomy, though they work manually (command line, rqt).  Have some names changed?   For example I see "UnstowRequest" in rqt but autonomy calls "Unstow".

Samahu commented:
> > If you start noticing too much "LCP error" message in the log, unfortunately these have happened after I synchronized this branch with _melodic-devel_ to resolve merge conflicts before actually merging.
> 
> The LCP error message are found in the branch before the synchronization with melodic-devel. I am not sure at what point of this branch/PR these started to happen. It wasn't much of a problem early on.
> 
> These are not necessarily an issue as we can silence them

@damianacatanoso I went back through the branch to find out at which point we started getting LCP error messages. What I have seen so far suggests that this happened after we merged https://github.com/nasa/ow_simulator/pull/50 (of https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-529 for Joint Constraint). I don't think the joint constraint has anything to do with the LCP error, but in this PR we also modified the PID values. So since you are already working on tuning PID values you maybe able to resolve this issue too!

Samahu commented:
> > When I launch atacama, the antenna is shaking in both Gazebo and Rviz. I presume this is a known and unrelated issue.
> 
> @Samahu I would try to merge the latest lander.xacro. This has new values for the antenna parameters

@kmdalal Damiana is referring to the open PR https://github.com/nasa/ow_simulator/pull/54 which should address the problem with the antenna. The only thing is that PR has more changes like modified effort and damping values. @damianacatanoso I will approve it for merge, but you might need to modify some of these values again as part of tuning PID values ticket.

Samahu commented:
> Rviz now keeps repeating the planning motion, which is pretty distracting. Is there a way to stop it or have it just do the motion once?

@mikedalal it is a simple rviz configuration, the image below shows the setting to disable it (Loop Animation)

![image](https://user-images.githubusercontent.com/606033/102802559-af819980-436b-11eb-947a-bef6aba51e6a.png)

Damiana brought this up last week, I told her I am fine either way. It is a simple question of what want to be the default setting.

damianacatanoso commented:
> > > If you start noticing too much "LCP error" message in the log, unfortunately these have happened after I synchronized this branch with _melodic-devel_ to resolve merge conflicts before actually merging.
> > 
> > 
> > The LCP error message are found in the branch before the synchronization with melodic-devel. I am not sure at what point of this branch/PR these started to happen. It wasn't much of a problem early on.
> > These are not necessarily an issue as we can silence them
> 
> @damianacatanoso I went back through the branch to find out at which point we started getting LCP error messages. What I have seen so far suggests that this happened after we merged #50 (of https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-529 for Joint Constraint). I don't think the joint constraint has anything to do with the LCP error, but in this PR we also modified the PID values. So since you are already working on tuning PID values you maybe able to resolve this issue too!

Thanks for looking into that. It sounds pretty strange to me that changing the PID values introduces the LCP error, since the latter relates to contact forces. The only thing I can think of is that wrong PID tuning causes less smooth joint motions, that causes a rough interaction between parts of the arm, which turns into high contact forces. However, observing the joint states I don't see any vibrations or oscillations. Anyway, I'll look into it.

damianacatanoso commented:
It turns out that damping=10 for the arm joints yields instability. I am re-tuning this value in the hope to eliminate the LCP error as well.

Samahu commented:
Revised joints damping values to 50 based on @damianacatanoso's recommendation

damianacatanoso commented:
While executing "Grind" I get this error:
[ERROR] [1608584008.916108656, 1916524963.980000019]: Controller is taking too long to execute trajectory (the expected upper bound for the trajectory execution was 1.404151 seconds). Stopping trajectory.    
Visually, the grinding trajectory has been successfully executed so I am not sure what this error implies.

Samahu commented:
> While executing "Grind" I get this error:
> [ERROR] [1608584008.916108656, 1916524963.980000019]: Controller is taking too long to execute trajectory (the expected upper bound for the trajectory execution was 1.404151 seconds). Stopping trajectory.
> Visually, the grinding trajectory has been successfully executed so I am not sure what this error implies.

Good catch Damiana, I think this corresponds to the **allowed_execution_duration_scaling** param within `trajectory_execution.launch` which is currently set to 1.2. I will increase the value to 2.0.

damianacatanoso commented:
- While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
[ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)

damianacatanoso commented:
- I got another error while grinding:
![image](https://user-images.githubusercontent.com/51930968/102821907-d2707580-438c-11eb-83b4-71d69392feea.png)

damianacatanoso commented:
- It seems Stow is not found by autonomy:
![image](https://user-images.githubusercontent.com/51930968/102822270-94c01c80-438d-11eb-84c8-8760cf154aaf.png)

Samahu commented:
> * I got another error while grinding:
>   ![image](https://user-images.githubusercontent.com/51930968/102821907-d2707580-438c-11eb-83b4-71d69392feea.png)

Should be addressed in the last commit.

damianacatanoso commented:
> * While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
>   [ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
>   ![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)

Just tested ReferenceMission1.plx again and this is the only error I see. All others are gone!

Samahu commented:
> > * While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
> >   [ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
> >   ![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)
> 
> Just tested ReferenceMission1.plx again and this is the only error I see. All others are gone!

Based on two isolated tests I couldn't generate the reported problem. Is anyone else beside Damiana observing the problem?

damianacatanoso commented:
> Based on two isolated tests I couldn't generate the reported problem. Is anyone else besides Damiana observing the problem?

If you want to reproduce the problem I'd run DigCircular alone multiple times, say 10 times

mogumbo commented:
> > Based on two isolated tests I couldn't generate the reported problem. Is anyone else besides Damiana observing the problem?
> 
> If you want to reproduce the problem I'd run DigCircular alone multiple times, say 10 times

I also see the digging motion high in the air about 25% of the time, but my error is different:
```
Dig Cicular arm activity started
[ INFO] [1608593060.411039225, 1916524987.467499971]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593060.411157127, 1916524987.467499971]: Planning attempt 1 of at most 1
[ INFO] [1608593060.411845424, 1916524987.467499971]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593060.412197273, 1916524987.467499971]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593060.430909326, 1916524987.480000019]: RRTConnect: Created 5 states (2 start + 3 goal)
[ INFO] [1608593060.430970792, 1916524987.480000019]: Solution found in 0.018958 seconds
[ INFO] [1608593060.477259485, 1916524987.532500029]: SimpleSetup: Path simplification took 0.046225 seconds and changed from 4 to 2 states
[ INFO] [1608593072.777669936, 1916524999.250000000]: Controller arm_controller successfully finished
[ INFO] [1608593072.792826683, 1916524999.259999990]: Completed trajectory execution with status SUCCEEDED ...
[ INFO] [1608593072.798484745, 1916524999.269999981]: Received event 'stop'
[ INFO] [1608593072.835444538, 1916524999.305000067]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593072.835520531, 1916524999.305000067]: Planning attempt 1 of at most 1
[ INFO] [1608593072.836005888, 1916524999.305000067]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593072.836284657, 1916524999.305000067]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593072.850592990, 1916524999.319999933]: RRTConnect: Created 4 states (2 start + 2 goal)
[ INFO] [1608593072.850644179, 1916524999.319999933]: Solution found in 0.014470 seconds
[ INFO] [1608593072.866137431, 1916524999.335000038]: SimpleSetup: Path simplification took 0.015445 seconds and changed from 3 to 2 states
[ INFO] [1608593080.131063256, 1916525006.497499943]: Controller arm_controller successfully finished
[ INFO] [1608593080.162990920, 1916525006.529999971]: Completed trajectory execution with status SUCCEEDED ...
[ INFO] [1608593080.169194931, 1916525006.535000086]: Received event 'stop'
[ INFO] [1608593080.200321025, 1916525006.565000057]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593080.200545741, 1916525006.565000057]: Planning attempt 1 of at most 1
[ INFO] [1608593080.201207740, 1916525006.565000057]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593080.201651771, 1916525006.565000057]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593080.215706700, 1916525006.577500105]: RRTConnect: Created 4 states (2 start + 2 goal)
[ INFO] [1608593080.215764036, 1916525006.577500105]: Solution found in 0.014204 seconds
[ INFO] [1608593080.257297598, 1916525006.612499952]: SimpleSetup: Path simplification took 0.041472 seconds and changed from 3 to 2 states
[ INFO] [1608593091.302864649, 1916525017.630000114]: Controller arm_controller successfully finished
[ INFO] [1608593091.337900011, 1916525017.664999962]: Completed trajectory execution with status SUCCEEDED ...
[ INFO] [1608593091.342832512, 1916525017.670000076]: Received event 'stop'
[ INFO] [1608593091.358108841, 1916525017.684999943]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593091.358231603, 1916525017.684999943]: Planning attempt 1 of at most 1
[ INFO] [1608593091.358773707, 1916525017.684999943]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593091.358986724, 1916525017.684999943]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593091.376760712, 1916525017.702500105]: RRTConnect: Created 4 states (2 start + 2 goal)
[ INFO] [1608593091.376833734, 1916525017.702500105]: Solution found in 0.017939 seconds
[ INFO] [1608593091.403584986, 1916525017.730000019]: SimpleSetup: Path simplification took 0.026673 seconds and changed from 3 to 2 states
[ INFO] [1608593097.038762867, 1916525023.289999962]: Controller arm_controller successfully finished
[ INFO] [1608593097.093522935, 1916525023.345000029]: Completed trajectory execution with status SUCCEEDED ...
[ INFO] [1608593097.101318982, 1916525023.352499962]: Received event 'stop'
[ INFO] [1608593097.133267008, 1916525023.384999990]: Planning request received for MoveGroup action. Forwarding to planning pipeline.
[ INFO] [1608593097.133884031, 1916525023.384999990]: Planner configuration 'limbs' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593097.134097792, 1916525023.384999990]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593097.220382725, 1916525023.454999924]: RRTConnect: Created 4 states (2 start + 2 goal)
[ INFO] [1608593097.221091858, 1916525023.457499981]: Solution found in 0.087086 seconds
[ INFO] [1608593097.263869172, 1916525023.497499943]: SimpleSetup: Path simplification took 0.038713 seconds and changed from 3 to 2 states
[ INFO] [1608593097.270676064, 1916525023.505000114]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593097.270970161, 1916525023.505000114]: Planning attempt 1 of at most 1
[ INFO] [1608593097.271840413, 1916525023.505000114]: Planner configuration 'limbs' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593097.272358083, 1916525023.505000114]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593097.551018476, 1916525023.777499914]: RRTConnect: Created 5 states (3 start + 2 goal)
[ INFO] [1608593097.551081595, 1916525023.777499914]: Solution found in 0.278885 seconds
[ INFO] [1608593097.581291568, 1916525023.809999943]: SimpleSetup: Path simplification took 0.019816 seconds and changed from 4 to 2 states
[ERROR] [1608593097.597187388, 1916525023.825000048]: 
Invalid Trajectory: start point deviates from current robot state more than 0.02
joint 'j_dist_pitch': expected: -1.10242, current: -1.12845
[ INFO] [1608593097.598018568, 1916525023.825000048]: ABORTED: Solution found but controller failed during execution
[ INFO] [1608593097.598225960, 1916525023.825000048]: Received event 'stop'
[ INFO] [1608593097.654703348, 1916525023.884999990]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
[ INFO] [1608593097.654796849, 1916525023.884999990]: Planning attempt 1 of at most 1
[ INFO] [1608593097.655435061, 1916525023.884999990]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
[ INFO] [1608593097.655689333, 1916525023.884999990]: RRTConnect: Starting planning with 1 states already in datastructure
[ INFO] [1608593097.673333522, 1916525023.902499914]: RRTConnect: Created 4 states (2 start + 2 goal)
[ INFO] [1608593097.673399205, 1916525023.902499914]: Solution found in 0.017817 seconds
[ INFO] [1608593097.721165393, 1916525023.952500105]: SimpleSetup: Path simplification took 0.047698 seconds and changed from 3 to 2 states
[ INFO] [1608593108.470982036, 1916525034.464999914]: Controller arm_controller successfully finished
[ INFO] [1608593108.471366979, 1916525034.464999914]: Completed trajectory execution with status SUCCEEDED ...
Dig Circular arm activity completed
[ INFO] [1608593108.483863022, 1916525034.477499962]: Received event 'stop'
```

kmdalal commented:
> > Rviz now keeps repeating the planning motion, which is pretty distracting. Is there a way to stop it or have it just do the motion once?
> 
> @mikedalal it is a simple rviz configuration, the image below shows the setting to disable it (Loop Animation)

Thanks, most helpful.  What do you think of making disabled the default?   Would this be easy?

Samahu commented:
> > > Rviz now keeps repeating the planning motion, which is pretty distracting. Is there a way to stop it or have it just do the motion once?
> > 
> > 
> > @mikedalal it is a simple rviz configuration, the image below shows the setting to disable it (Loop Animation)
> 
> Thanks, most helpful. What do you think of making disabled the default? Would this be easy?

Yes, very easy. I will disable it.

Samahu commented:
> > > Based on two isolated tests I couldn't generate the reported problem. Is anyone else besides Damiana observing the problem?
> > 
> > 
> > If you want to reproduce the problem I'd run DigCircular alone multiple times, say 10 times
> 
> I also see the digging motion high in the air about 25% of the time, but my error is different:
> 
> ```
> Dig Cicular arm activity started
> [ INFO] [1608593060.411039225, 1916524987.467499971]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593060.411157127, 1916524987.467499971]: Planning attempt 1 of at most 1
> [ INFO] [1608593060.411845424, 1916524987.467499971]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593060.412197273, 1916524987.467499971]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593060.430909326, 1916524987.480000019]: RRTConnect: Created 5 states (2 start + 3 goal)
> [ INFO] [1608593060.430970792, 1916524987.480000019]: Solution found in 0.018958 seconds
> [ INFO] [1608593060.477259485, 1916524987.532500029]: SimpleSetup: Path simplification took 0.046225 seconds and changed from 4 to 2 states
> [ INFO] [1608593072.777669936, 1916524999.250000000]: Controller arm_controller successfully finished
> [ INFO] [1608593072.792826683, 1916524999.259999990]: Completed trajectory execution with status SUCCEEDED ...
> [ INFO] [1608593072.798484745, 1916524999.269999981]: Received event 'stop'
> [ INFO] [1608593072.835444538, 1916524999.305000067]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593072.835520531, 1916524999.305000067]: Planning attempt 1 of at most 1
> [ INFO] [1608593072.836005888, 1916524999.305000067]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593072.836284657, 1916524999.305000067]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593072.850592990, 1916524999.319999933]: RRTConnect: Created 4 states (2 start + 2 goal)
> [ INFO] [1608593072.850644179, 1916524999.319999933]: Solution found in 0.014470 seconds
> [ INFO] [1608593072.866137431, 1916524999.335000038]: SimpleSetup: Path simplification took 0.015445 seconds and changed from 3 to 2 states
> [ INFO] [1608593080.131063256, 1916525006.497499943]: Controller arm_controller successfully finished
> [ INFO] [1608593080.162990920, 1916525006.529999971]: Completed trajectory execution with status SUCCEEDED ...
> [ INFO] [1608593080.169194931, 1916525006.535000086]: Received event 'stop'
> [ INFO] [1608593080.200321025, 1916525006.565000057]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593080.200545741, 1916525006.565000057]: Planning attempt 1 of at most 1
> [ INFO] [1608593080.201207740, 1916525006.565000057]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593080.201651771, 1916525006.565000057]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593080.215706700, 1916525006.577500105]: RRTConnect: Created 4 states (2 start + 2 goal)
> [ INFO] [1608593080.215764036, 1916525006.577500105]: Solution found in 0.014204 seconds
> [ INFO] [1608593080.257297598, 1916525006.612499952]: SimpleSetup: Path simplification took 0.041472 seconds and changed from 3 to 2 states
> [ INFO] [1608593091.302864649, 1916525017.630000114]: Controller arm_controller successfully finished
> [ INFO] [1608593091.337900011, 1916525017.664999962]: Completed trajectory execution with status SUCCEEDED ...
> [ INFO] [1608593091.342832512, 1916525017.670000076]: Received event 'stop'
> [ INFO] [1608593091.358108841, 1916525017.684999943]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593091.358231603, 1916525017.684999943]: Planning attempt 1 of at most 1
> [ INFO] [1608593091.358773707, 1916525017.684999943]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593091.358986724, 1916525017.684999943]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593091.376760712, 1916525017.702500105]: RRTConnect: Created 4 states (2 start + 2 goal)
> [ INFO] [1608593091.376833734, 1916525017.702500105]: Solution found in 0.017939 seconds
> [ INFO] [1608593091.403584986, 1916525017.730000019]: SimpleSetup: Path simplification took 0.026673 seconds and changed from 3 to 2 states
> [ INFO] [1608593097.038762867, 1916525023.289999962]: Controller arm_controller successfully finished
> [ INFO] [1608593097.093522935, 1916525023.345000029]: Completed trajectory execution with status SUCCEEDED ...
> [ INFO] [1608593097.101318982, 1916525023.352499962]: Received event 'stop'
> [ INFO] [1608593097.133267008, 1916525023.384999990]: Planning request received for MoveGroup action. Forwarding to planning pipeline.
> [ INFO] [1608593097.133884031, 1916525023.384999990]: Planner configuration 'limbs' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593097.134097792, 1916525023.384999990]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593097.220382725, 1916525023.454999924]: RRTConnect: Created 4 states (2 start + 2 goal)
> [ INFO] [1608593097.221091858, 1916525023.457499981]: Solution found in 0.087086 seconds
> [ INFO] [1608593097.263869172, 1916525023.497499943]: SimpleSetup: Path simplification took 0.038713 seconds and changed from 3 to 2 states
> [ INFO] [1608593097.270676064, 1916525023.505000114]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593097.270970161, 1916525023.505000114]: Planning attempt 1 of at most 1
> [ INFO] [1608593097.271840413, 1916525023.505000114]: Planner configuration 'limbs' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593097.272358083, 1916525023.505000114]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593097.551018476, 1916525023.777499914]: RRTConnect: Created 5 states (3 start + 2 goal)
> [ INFO] [1608593097.551081595, 1916525023.777499914]: Solution found in 0.278885 seconds
> [ INFO] [1608593097.581291568, 1916525023.809999943]: SimpleSetup: Path simplification took 0.019816 seconds and changed from 4 to 2 states
> [ERROR] [1608593097.597187388, 1916525023.825000048]: 
> Invalid Trajectory: start point deviates from current robot state more than 0.02
> joint 'j_dist_pitch': expected: -1.10242, current: -1.12845
> [ INFO] [1608593097.598018568, 1916525023.825000048]: ABORTED: Solution found but controller failed during execution
> [ INFO] [1608593097.598225960, 1916525023.825000048]: Received event 'stop'
> [ INFO] [1608593097.654703348, 1916525023.884999990]: Combined planning and execution request received for MoveGroup action. Forwarding to planning and execution pipeline.
> [ INFO] [1608593097.654796849, 1916525023.884999990]: Planning attempt 1 of at most 1
> [ INFO] [1608593097.655435061, 1916525023.884999990]: Planner configuration 'arm' will use planner 'geometric::RRTConnect'. Additional configuration parameters will be set when the planner is constructed.
> [ INFO] [1608593097.655689333, 1916525023.884999990]: RRTConnect: Starting planning with 1 states already in datastructure
> [ INFO] [1608593097.673333522, 1916525023.902499914]: RRTConnect: Created 4 states (2 start + 2 goal)
> [ INFO] [1608593097.673399205, 1916525023.902499914]: Solution found in 0.017817 seconds
> [ INFO] [1608593097.721165393, 1916525023.952500105]: SimpleSetup: Path simplification took 0.047698 seconds and changed from 3 to 2 states
> [ INFO] [1608593108.470982036, 1916525034.464999914]: Controller arm_controller successfully finished
> [ INFO] [1608593108.471366979, 1916525034.464999914]: Completed trajectory execution with status SUCCEEDED ...
> Dig Circular arm activity completed
> [ INFO] [1608593108.483863022, 1916525034.477499962]: Received event 'stop'
> ```

I can resolve this issue easily by increasing the `trajectory_execution/allowed_start_tolerance` again but I think this problem and some of the others reported recently are side effects of an insufficient PID values.

damianacatanoso commented:
> I can resolve this issue easily by increasing the `trajectory_execution/allowed_start_tolerance` again but I think this problem and some of the others reported recently are side effects of an insufficient PID values.

Maybe I am doing something wrong with the tuning process then. I suggest we have a call and we discuss this.

damianacatanoso commented:
> > > * While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
> > >   [ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
> > >   ![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)
> > 
> > 
> > Just tested ReferenceMission1.plx again and this is the only error I see. All others are gone!
> 
> Based on two isolated tests I couldn't generate the reported problem. Is anyone else besides Damiana observing the problem?

I was testing ReferenceMission1 and I got this error again

Samahu commented:
> > * While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
> >   [ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
> >   ![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)
> 
> Just tested ReferenceMission1.plx again and this is the only error I see. All others are gone!



> > > > * While executing DigCircular, the scoop doesn't approach the ground and performs the circular motion in the air. This is the error:
> > > >   [ERROR] [1608584309.643541706, 1916525264.535000086]: RRTConnect: Unable to sample any valid states for goal tree
> > > >   ![image](https://user-images.githubusercontent.com/51930968/102821803-92a98e00-438c-11eb-9292-ef39618f8fad.png)
> > > 
> > > 
> > > Just tested ReferenceMission1.plx again and this is the only error I see. All others are gone!
> > 
> > 
> > Based on two isolated tests I couldn't generate the reported problem. Is anyone else besides Damiana observing the problem?
> 
> I was testing ReferenceMission1 and I got this error again

I created a ticket to track this issue https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-569
This seem to happen when the planner is given an unreachable target.

kmdalal reviewed (approved):
After running ReferenceMission1 several times and doing a fair amount of manual testing, I'm not seeing any significant problems -- just occasional LCP errors (which where there before) and other misc warnings.    

My review of the file changes was pretty superficial, as I don't know many of the ROS files very well (yaml, rviz, etc).

jfugate1 reviewed (approved):
A couple files are missing new line characters at the end of the file. Shouldn't be an issue though.
- config/kinematics.yaml 
- config/joint_limits.yaml 
- config/ompl_planning.yaml
- launch/planning_pipeline.launch.xml
